doc/developer/design/20210831_correctness.md

```diff
@@ -0,0 +1,124 @@
+# <Insert Name of Feature>
+
+## Summary
+
+A description of the correctness properties Materialize provides and the terms and concepts we use to describe them.
+
+## Goals
+
+The main goal is to lay the foundation for ongoing work to clearly articulate correctness properties of Materialize.
+This includes the introduction of concepts that underlie correctess, the components that will enforce them, and how they can communicate clearly.
+
+## Non-Goals
+
+It is not a goal to prescribe fine detail in how each of the components should behave, or which should have certain responsibilities.
+It is not a goal to prescribe which behavior Materialize should support, but only how to speak about a "correct" subset of it.
+It is not a goal to advocate (much) for DEFINITE over NON-VOLATILE.
+
+## Description
+
+Materialize means to provide several correctness guarantees, which manifest as various consistency and durability properties.
+At the heart of this is the idea that the ordering of events can be explicitly resolved and recorded, after which queries and views over the data become a matter of correct evaluation on completed subsets of time.
+The guarantees of Materialize derive from the durability of these recorded data, and the determinism of views over them.
+
+The primary term introduced here is "definite" to modify a collection of data.
+A collection is definite if it is uniquely named or defined such that anyone with the same identification will see the same collection of data at the same logical times.
+A definite collection has had its updates put in a specific order, by way of the timestamps on updates.
+A handle to a collection can be "definite from `since` until `upper`" (names pending) if it reflects the definite results for times greater or equal to `since` but not greater or equal to `upper`.
+Among the correctness guarantees of Materialize, a main one will be statements of definiteness over intervals of times for collections it maintains.
+
+### Components
+
+There are three components to discuss: storage, compute, and coordination.
+
+The **storage layer** (`persist`, plus potentially others) presents streams of updates to collections to other parts of the system (primary the compute layer).
+One correctness goal of the `persist` module is to make as many update streams as possible "definite", even if the raw data are indefinite (e.g. perhaps they contain no timestamps, which must be added).
```

Comment:
"as many" or "all"?

Reply:
I think just "as many", in that per the VOLATILE design doc, we 100% expect volatile sources and it is a non-starter to proscribe them entirely. I don't know if the storage layer ends up having any opinion on them (does it provide them, without guarantees, or does it just excuse itself from their definition entirely), but I think the feeling is they should be allowed to exist (although, potentially dramatically limiting what grown-up features you can use with them).

Reply:
Understood!